src/backend/distributed/planner/distributed_planner.c

```diff
@@ -1112,23 +1112,49 @@ FinalizeDistributedPlan(DistributedPlan *plan, Query *originalQuery)
 static void
 RecordSubPlansUsedInPlan(DistributedPlan *plan, Query *originalQuery)
 {
-	/* first, get all the subplans in the query */
-	plan->usedSubPlanNodeList = FindSubPlansUsedInNode((Node *) originalQuery);
+	Node *havingQual = originalQuery->havingQual;
+
+	/* temporarily set to NULL, we're going to restore before the function returns */
+	originalQuery->havingQual = NULL;
 
 	/*
-	 * Later, remove the subplans used in the HAVING clause, because they
-	 * are only required in the coordinator. Including them in the
-	 * usedSubPlanNodeList prevents the intermediate results to be sent to the
-	 * coordinator only.
+	 * Mark the subplans as needed on remote side. Note that this decision is revisited
+	 * on execution, when the query only consists of intermediate results.
 	 */
+	List *subplansExceptHaving = FindSubPlansUsedInNode((Node *) originalQuery);
+	UpdateUsedPlanListLocation(subplansExceptHaving, SUBPLAN_ACCESS_REMOTE);
+
+	/* do the same for HAVING part of the query */
+	List *subplansInHaving = NIL;
 	if (originalQuery->hasSubLinks &&
-		FindNodeCheck(originalQuery->havingQual, IsNodeSubquery))
+		FindNodeCheck(havingQual, IsNodeSubquery))
 	{
-		List *subplansInHaving = FindSubPlansUsedInNode(originalQuery->havingQual);
-
-		plan->usedSubPlanNodeList =
-			list_difference(plan->usedSubPlanNodeList, subplansInHaving);
+		subplansInHaving = FindSubPlansUsedInNode(havingQual);
+		if (plan->masterQuery)
```

Comment:
I keep coming back to this, and it is probably because my lack of knowledge of intimate details in this area, but could there be a case where we have a `masterQuery` and a subplan only used in a having clause where we might also need to send it to the remote workers? I guess not, and I guess we have a test that covers it somewhere.

But next question is, will this always stay like this?

Are there situations where we push having clauses down to the worker (eg. we need the subplan on the worker) but still have a `masterQuery`. And if we don't have such cases at the moment, when might we introduce the support for sql that might run into this situation.

Looking at https://jvns.ca/blog/2019/10/03/sql-queries-don-t-start-with-select/ it seems this could be introduced with a Having on an aggregate which is grouped by the distribution column and having an `ORDER BY`.

From looking at the diff we do not have added any tests that have both `HAVING` and `ORDER BY`. Can we add a test case for Having on an aggregate on a grouping that contains the distribution column with an Order By.

Reply:
I've been doing some testing with this:
```sql
WITH cte_1 AS MATERIALIZED (SELECT max(value) FROM table_1)
SELECT key
FROM table_2
GROUP BY key
HAVING max(value) >
  (SELECT max FROM cte_1)
ORDER BY key ASC;
```

It seems we do not push down the having clause, so this makes sense now.
I think adding this test to the test suite will help us once we decide it is a good idea to push having for aggregates on distribution columns.

Besides that this is ready to land.

Reply:
We've the same query in the tests, actually the first one :) 

The trick in the test file to make all the CTEs `MATERIALIZED ` is the following in case you're not familiar with the GUC, it prevents all CTE inlining for pg 11 and pg 12 on Citus 
```SQL
-- prevent PG 11 - PG 12 outputs to diverge
-- and have a lot more CTEs recursively planned for the
-- sake of increasing the test coverage
SET citus.enable_cte_inlining TO false;
```